File: leetcode/python/001_two_sum.py
# -*- coding: utf-8 -*-
# @Author:             dugufei (dugufei@bjtu.edu.cn)
# @Date:               2022-03-15 23:59:49
# @Last Modified by:   dugufei (dugufei@bjtu.edu.cn)
# @Last Modified time: 2022-03-16 00:01:57

class Solution:
    def twoSum(self, nums: list[int], target: int) -> list[int]:
        sub_dict = {}
        for indx in range(len(nums)):
            sub = target - nums[indx]
            if sub in sub_dict.keys():
                return sub_dict[sub], indx
            sub_dict[nums[indx]] = indx
        return -1,-1


if __name__ == "__main__":
    solution = Solution()
    print(solution.twoSum([2,7,11,15], 9))

File: leetcode/00001_two_sum/001_two_sum.cpp
/*
 * @Author:             dugufei (dugufei@bjtu.edu.cn)
 * @Date:               2021-05-18 20:33:02
 * @Last Modified by:   dugufei (dugufei@bjtu.edu.cn)
 * @Last Modified time: 2022-03-16 00:02:17
 */

#include <iostream>
#include <vector>
#include <unordered_map>

class Solution {
public:
    std::vector<int> twoSum(std::vector<int>& nums, int target) {
        std::unordered_map<int, int>  sub_map;
        for (int indx = 0; indx < nums.size(); ++indx) {
            int sub = target - nums[indx];
            if (sub_map.find(sub) != sub_map.end()) {
                return {sub_map[sub], indx};
            }
            sub_map[nums[indx]] = indx;
        }
        return {-1, -1};
    }
};

int main() {
    std::vector<int> nums{2,7,11,15};
    Solution solution;
    auto tag = solution.twoSum(nums, 9);
    for (auto const& iter : tag) {
        std::cout << iter << std::endl;
    }
    return 0;
}
